[CSS] <iframe> should be scaled by the used zoom

mdubet opened the pull request:
#### 24964b8e4316b1c52177852ec082d1f109c1ca68
<pre>
[CSS] &lt;iframe&gt; should be scaled by the used zoom
<a href="https://bugs.webkit.org/show_bug.cgi?id=301148">https://bugs.webkit.org/show_bug.cgi?id=301148</a>

Reviewed by Sammy Gill.

We take into account the usedZoom() of the owning &lt;iframe&gt; for content inside it
via frameScaleFactor().

* LayoutTests/TestExpectations:

The site-isolation/page-zoom test need further investigation

* LayoutTests/fast/transforms/bounding-rect-zoom-expected.txt:
* LayoutTests/fast/transforms/bounding-rect-zoom.html:

Add some more cases with &lt;iframe&gt; and pixelDeviceRatio

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic.html:

Fix flakyness when the event listener is set too late

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-2-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-huge-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-override-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-percentage-size-expected.txt:
* LayoutTests/svg/zoom/page/zoom-svg-as-object-expected.txt:

All those tests now correctly render the circle taking into account the zoom

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::devicePixelRatio const):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::frameScaleFactor const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::reset):
(WebCore::LocalFrameView::willDoLayout):
(WebCore::LocalFrameView::scheduleResizeEventIfNeeded):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::styleDidChange):

Canonical link: <a href="https://commits.webkit.org/302097@main">https://commits.webkit.org/302097@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3db069d8589e10f02bbd0abf01159f5d5c8ecc17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135389 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79530 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3fe73dea-dd5f-4ff7-a5d7-0ab29f7a0daa) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129890 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97456 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65347 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7ac2a4aa-3984-4030-9e6e-8283386962b4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130966 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114681 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78025 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b45b238d-1d8e-4e99-b797-56ee1afbc68b) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32786 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78698 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/108479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33266 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137872 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/157 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/151 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/182 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111031 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105721 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26946 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/120 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29580 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52325 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/203 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/61663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/122 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/199 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/163 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
Do we pass the invalidation test? https://github.com/web-platform-tests/wpt/blob/master/css/css-viewport/zoom/zoom-iframe-dynamic.html

sammygill commented on Source/WebCore/rendering/RenderWidget.cpp:
```diff
@@ -241,6 +241,18 @@ void RenderWidget::styleDidChange(StyleDifference diff, const RenderStyle* oldSt
+    if (oldStyle && oldStyle->zoom() != style().zoom()) {
+        if (auto* frameView = dynamicDowncast<LocalFrameView>(m_widget.get())) {
+            if (auto* frame = &frameView->frame()) {
+                frame->deviceOrPageScaleFactorChanged();
+                if (auto* view = frameView)
+                    view->scheduleResizeEventIfNeeded();
+            }
+        }
```
Maybe do some early returns here to reduce the number of nested scopes?

mdubet replied:
Cleaned (some of those scopes were actually useless..)

vitorroriz commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1124,10 +1124,20 @@ float LocalFrame::frameScaleFactor() const
+    // https://github.com/w3c/csswg-drafts/issues/9644
+    // Check if this frame's owner element (iframe) has CSS zoom applied.
+    if (!isMainFrame()) {
+        if (RefPtr ownerElement = this->ownerElement()) {
+            if (auto* ownerRenderer = ownerElement->renderer())
+                return ownerRenderer->style().usedZoom();
+        }
+        return 1;
```
nit: maybe it would be a bit cleaner to isolate this on its own private function member

vitorroriz commented on LayoutTests/TestExpectations:
```diff
@@ -7381,6 +7381,7 @@ imported/w3c/web-platform-tests/dom/events/scrolling/wheel-event-transactions-ta
 
 imported/w3c/web-platform-tests/css/css-color/deprecated-sameas-002.html [ ImageOnlyFailure ]
 
+http/tests/site-isolation/page-zoom.html [ Skip ]
```
why do we skip this? is this something you want to investigate on a separated PR? If so can you add the bugzilla link here?

mdubet replied:
added

mdubet committed:
[CSS] <iframe> should be scaled by the used zoom
https://bugs.webkit.org/show_bug.cgi?id=301148

Reviewed by Sammy Gill.

We take into account the usedZoom() of the owning <iframe> for content inside it
via frameScaleFactor().

* LayoutTests/TestExpectations:

The site-isolation/page-zoom test need further investigation

* LayoutTests/fast/transforms/bounding-rect-zoom-expected.txt:
* LayoutTests/fast/transforms/bounding-rect-zoom.html:

Add some more cases with <iframe> and pixelDeviceRatio

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic.html:

Fix flakyness when the event listener is set too late

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-2-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-huge-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-override-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-percentage-size-expected.txt:
* LayoutTests/svg/zoom/page/zoom-svg-as-object-expected.txt:

All those tests now correctly render the circle taking into account the zoom

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::devicePixelRatio const):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::frameScaleFactor const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::reset):
(WebCore::LocalFrameView::willDoLayout):
(WebCore::LocalFrameView::scheduleResizeEventIfNeeded):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::styleDidChange):

Canonical link: https://commits.webkit.org/302097@main
